python_modules/dagster/dagster/_config/structured_config/__init__.py

```diff
@@ -168,13 +180,154 @@ def _curry_config_schema(schema_field: Field, data: Any) -> IDefinitionConfigSch
+    ):
+        self._resource = resource
+        self._resource_id_to_key_mapping = resource_id_to_key_mapping
+
+        ResourceDefinition.__init__(
+            self,
+            resource_fn=self.setup_context_resources_and_call,
+            config_schema=resource.config_schema,
+            description=resource.description,
+            version=resource.version,
+        )
+
+    def setup_context_resources_and_call(self, context: InitResourceContext):
+        """
+        Wrapper around the wrapped resource's resource_fn which sets up the context.resources
+        to include resources by their ID, and then calls the nested resource's resource_fn.
+        """
+        for resource_id, resource_key in self._resource_id_to_key_mapping.items():
+            # Wrapped resource only knows about its nested resources by ID, so we need to
+            # set up the context.resources to include resources by their ID as well.
+            setattr(
+                context.resources,
+                f"id_{resource_id}",
+                getattr(context.resources, resource_key, None),
+            )
+
+        if has_at_least_one_parameter(self._resource.resource_fn):
+            return self._resource.resource_fn(context)
+        else:
+            return cast(ResourceFunctionWithoutContext, self._resource.resource_fn)()
+
+    @property
+    def required_resource_keys(self) -> AbstractSet[str]:
+        return _resolve_required_resource_keys_for_resource(
+            self._resource, self._resource_id_to_key_mapping
+        )
+
+
+class IOManagerWithKeyMapping(ResourceWithKeyMapping, IOManagerDefinition):
+    """
+    Version of ResourceWithKeyMapping wrapper that also implements IOManagerDefinition.
+    """
+
+    def __init__(
+        self, resource: ResourceDefinition, resource_id_to_key_mapping: Dict[ResourceId, str]
+    ):
+        ResourceWithKeyMapping.__init__(self, resource, resource_id_to_key_mapping)
+        IOManagerDefinition.__init__(
+            self, resource_fn=self.resource_fn, config_schema=resource.config_schema
+        )
+
+
+def attach_resource_id_to_key_mapping(
+    resource_def: Any, resource_id_to_key_mapping: Dict[ResourceId, str]
+) -> Any:
+    if isinstance(resource_def, ResourceDefinition):
+        return (
+            IOManagerWithKeyMapping(resource_def, resource_id_to_key_mapping)
+            if isinstance(resource_def, IOManagerDefinition)
+            else ResourceWithKeyMapping(resource_def, resource_id_to_key_mapping)
+        )
+    return resource_def
+
+
+def _get_resource_by_id(context: InitResourceContext, resource_id: ResourceId) -> Any:
+    """
+    Retrieves a resource's value from the resource context object from the ResourceDefinition's ID.
+    This is used to retrieve the value of nested resources which are not fully configured, since the
+    resource key is not known by the nested resource's parent.
+    """
+    return getattr(context.resources, f"id_{resource_id}", None)
```

Comment:
we should model this with a data structure somewhere, not mangle strings and monkeypatch

Reply:
Attached to a new `InitResourceContext` subclass, `InitResourceContextWithKeyMapping`, which includes the id-to-resource mapping.